File: frontend/lib/api/documents.ts
/**
 * 文档 API
 */

import { apiGet, apiPost, apiPut, apiDelete, getApiBaseUrl } from "@/lib/api/client";

export type DocumentStatus = "pending" | "processing" | "completed" | "failed";

export interface Document {
  id: string;
  title: string;
  description: string | null;
  filename: string;
  file_path: string;
  file_size: number;
  file_type: string;
  content: string | null;
  status: DocumentStatus;
  error_message: string | null;
  category: string | null;
  tags: string[];
  folder_id: string | null;
  metadata: Record<string, unknown>;
  // 技能转化相关
  skill_id: string | null;
  is_converted: boolean;
  converted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentFolder {
  id: string;
  name: string;
  description: string | null;
  parent_id: string | null;
  sort_order: number;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface DocumentListResponse {
  items: Document[];
  total: number;
  page: number;
  page_size: number;
}

export interface DocumentFolderCreate {
  name: string;
  description?: string;
  parent_id?: string;
}

export interface DocumentFolderUpdate {
  name?: string;
  description?: string;
  parent_id?: string;
  sort_order?: number;
}

export interface DocumentUpdate {
  title?: string;
  description?: string;
  folder_id?: string | null;
  category?: string;
  tags?: string[];
}

// 文件夹 API
export function listFolders(parentId?: string): Promise<DocumentFolder[]> {
  const query = parentId ? `?parent_id=${parentId}` : "";
  return apiGet<DocumentFolder[]>(`/documents/folders${query}`);
}

export function getFolder(folderId: string): Promise<DocumentFolder> {
  return apiGet<DocumentFolder>(`/documents/folders/${folderId}`);
}

export function createFolder(data: DocumentFolderCreate): Promise<DocumentFolder> {
  return apiPost<DocumentFolder>("/documents/folders", data);
}

export function updateFolder(folderId: string, data: DocumentFolderUpdate): Promise<DocumentFolder> {
  return apiPut<DocumentFolder>(`/documents/folders/${folderId}`, data);
}

export function deleteFolder(folderId: string): Promise<{ success: boolean }> {
  return apiDelete<{ success: boolean }>(`/documents/folders/${folderId}`);
}

// 文档 API
export function listDocuments(params?: {
  folder_id?: string;
  category?: string;
  status?: DocumentStatus;
  is_converted?: boolean;
  page?: number;
  page_size?: number;
}): Promise<DocumentListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.folder_id) searchParams.set("folder_id", params.folder_id);
  if (params?.category) searchParams.set("category", params.category);
  if (params?.status) searchParams.set("status", params.status);
  if (params?.is_converted !== undefined) searchParams.set("is_converted", String(params.is_converted));
  if (params?.page) searchParams.set("page", String(params.page));
  if (params?.page_size) searchParams.set("page_size", String(params.page_size));
  
  const query = searchParams.toString();
  return apiGet<DocumentListResponse>(`/documents${query ? `?${query}` : ""}`);
}

export function getDocument(documentId: string): Promise<Document> {
  return apiGet<Document>(`/documents/${documentId}`);
}

export async function uploadDocument(
  file: File,
  options?: { title?: string; folder_id?: string }
): Promise<Document> {
  const formData = new FormData();
  formData.append("file", file);
  if (options?.title) formData.append("title", options.title);
  if (options?.folder_id) formData.append("folder_id", options.folder_id);

  const response = await fetch(`${getApiBaseUrl()}/documents/upload`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: "上传失败" }));
    throw new Error(error.detail || "上传失败");
  }

  return response.json();
}

export function updateDocument(documentId: string, data: DocumentUpdate): Promise<Document> {
  return apiPut<Document>(`/documents/${documentId}`, data);
}

export function deleteDocument(documentId: string): Promise<{ success: boolean }> {
  return apiDelete<{ success: boolean }>(`/documents/${documentId}`);
}

export function moveDocument(documentId: string, folderId: string | null): Promise<{ success: boolean }> {
  return apiPost<{ success: boolean }>(`/documents/${documentId}/move`, { folder_id: folderId });
}

export function searchDocuments(query: string, limit?: number): Promise<{ items: Document[]; total: number }> {
  const params = new URLSearchParams({ q: query });
  if (limit) params.set("limit", String(limit));
  return apiGet<{ items: Document[]; total: number }>(`/documents/search?${params}`);
}

export interface ConvertToSkillResult {
  success: boolean;
  skill: {
    id: string;
    name: string;
    description: string;
    category: string;
    trigger_keywords: string[];
  } | null;
  analysis: {
    doc_type: string;
    word_count: number;
    complexity: string;
    concepts: string[];
  } | null;
}

export function convertToSkill(
  documentId: string,
  options?: { use_llm?: boolean; folder_id?: string }
): Promise<ConvertToSkillResult> {
  const params = new URLSearchParams();
  if (options?.use_llm !== undefined) params.set("use_llm", String(options.use_llm));
  if (options?.folder_id) params.set("folder_id", options.folder_id);
  const query = params.toString();
  return apiPost<ConvertToSkillResult>(
    `/documents/${documentId}/convert-to-skill${query ? `?${query}` : ""}`,
    {}
  );
}

File: backend/app/routers/documents.py
"""文档管理路由"""

import os
import uuid
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.core.database import get_db
from app.models.document import DocumentStatus
from app.schemas.document import (
    DocumentCreate,
    DocumentUpdate,
    DocumentResponse,
    DocumentListResponse,
    DocumentFolderCreate,
    DocumentFolderUpdate,
    DocumentFolderResponse,
)
from app.services.document import DocumentService
from app.services.document_to_skill import DocumentToSkillService, ConvertOptions

router = APIRouter(prefix="/documents", tags=["documents"])


@router.post("/folders", response_model=DocumentFolderResponse)
async def create_folder(
    data: DocumentFolderCreate,
    db: AsyncSession = Depends(get_db),
):
    """创建文件夹"""
    service = DocumentService(db)
    folder = await service.create_folder(data)
    return folder


@router.get("/folders", response_model=list[DocumentFolderResponse])
async def list_folders(
    parent_id: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """列出文件夹"""
    service = DocumentService(db)
    folders = await service.list_folders(parent_id)
    return folders


@router.get("/folders/{folder_id}", response_model=DocumentFolderResponse)
async def get_folder(
    folder_id: str,
    db: AsyncSession = Depends(get_db),
):
    """获取文件夹"""
    service = DocumentService(db)
    folder = await service.get_folder(folder_id)
    if not folder:
        raise HTTPException(status_code=404, detail="文件夹不存在")
    return folder


@router.put("/folders/{folder_id}", response_model=DocumentFolderResponse)
async def update_folder(
    folder_id: str,
    data: DocumentFolderUpdate,
    db: AsyncSession = Depends(get_db),
):
    """更新文件夹"""
    service = DocumentService(db)
    folder = await service.update_folder(folder_id, data)
    if not folder:
        raise HTTPException(status_code=404, detail="文件夹不存在")
    return folder


@router.delete("/folders/{folder_id}")
async def delete_folder(
    folder_id: str,
    db: AsyncSession = Depends(get_db),
):
    """删除文件夹"""
    service = DocumentService(db)
    success = await service.delete_folder(folder_id)
    if not success:
        raise HTTPException(status_code=404, detail="文件夹不存在或无法删除")
    return {"success": True}


@router.post("/upload", response_model=DocumentResponse)
async def upload_document(
    file: UploadFile = File(...),
    title: str | None = None,
    folder_id: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """上传文档"""
    # 确保上传目录存在
    upload_dir = settings.ensure_data_dir() / "uploads"
    upload_dir.mkdir(exist_ok=True)

    # 生成文件路径
    file_ext = Path(file.filename).suffix if file.filename else ""
    file_id = str(uuid.uuid4())
    file_path = upload_dir / f"{file_id}{file_ext}"

    # 保存文件
    content = await file.read()
    with open(file_path, "wb") as f:
        f.write(content)

    # 提取文本内容（简单实现，仅支持文本文件）
    text_content = None
    if file_ext.lower() in [".txt", ".md", ".markdown"]:
        try:
            text_content = content.decode("utf-8")
        except Exception:
            pass

    # 创建文档记录
    service = DocumentService(db)
    data = DocumentCreate(
        title=title or file.filename or "未命名文档",
        folder_id=folder_id,
    )
    document = await service.create_document(
        data=data,
        filename=file.filename or "unknown",
        file_path=str(file_path),
        file_size=len(content),
        file_type=file_ext.lstrip(".") or "unknown",
        content=text_content,
    )
    return document


@router.get("", response_model=DocumentListResponse)
async def list_documents(
    folder_id: str | None = None,
    category: str | None = None,
    status: DocumentStatus | None = None,
    is_converted: bool | None = None,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    """列出文档
    
    Args:
        is_converted: 过滤是否已转化为技能（True=已转化, False=未转化, None=全部）
    """
    service = DocumentService(db)
    documents, total = await service.list_documents(
        folder_id=folder_id,
        category=category,
        status=status,
        is_converted=is_converted,
        page=page,
        page_size=page_size,
    )
    return DocumentListResponse(
        items=documents,
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get("/search")
async def search_documents(
    q: str = Query(..., min_length=1),
    limit: int = Query(default=20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    """搜索文档"""
    service = DocumentService(db)
    documents = await service.search_documents(q, limit)
    return {"items": documents, "total": len(documents)}


@router.get("/{document_id}", response_model=DocumentResponse)
async def get_document(
    document_id: str,
    db: AsyncSession = Depends(get_db),
):
    """获取文档详情"""
    service = DocumentService(db)
    document = await service.get_document(document_id)
    if not document:
        raise HTTPException(status_code=404, detail="文档不存在")
    return document


@router.put("/{document_id}", response_model=DocumentResponse)
async def update_document(
    document_id: str,
    data: DocumentUpdate,
    db: AsyncSession = Depends(get_db),
):
    """更新文档"""
    service = DocumentService(db)
    document = await service.update_document(document_id, data)
    if not document:
        raise HTTPException(status_code=404, detail="文档不存在")
    return document


@router.delete("/{document_id}")
async def delete_document(
    document_id: str,
    db: AsyncSession = Depends(get_db),
):
    """删除文档"""
    service = DocumentService(db)
    success = await service.delete_document(document_id)
    if not success:
        raise HTTPException(status_code=404, detail="文档不存在")
    return {"success": True}


@router.post("/{document_id}/move")
async def move_document(
    document_id: str,
    folder_id: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """移动文档到指定文件夹"""
    service = DocumentService(db)
    document = await service.move_document(document_id, folder_id)
    if not document:
        raise HTTPException(status_code=404, detail="文档不存在")
    return {"success": True, "folder_id": folder_id}


@router.post("/{document_id}/convert-to-skill")
async def convert_to_skill(
    document_id: str,
    use_llm: bool = True,
    auto_activate: bool = True,
    folder_id: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """将文档转换为 Skill
    
    Args:
        document_id: 文档 ID
        use_llm: 是否使用 LLM 生成（默认 True）
        auto_activate: 是否自动激活 Skill（默认 True）
        folder_id: 目标文件夹 ID
    
    Returns:
        转换结果，包含生成的 Skill 信息
    """
    options = ConvertOptions(
        use_llm=use_llm,
        auto_activate=auto_activate,
        folder_id=folder_id,
    )
    
    service = DocumentToSkillService(db)
    result = await service.convert(document_id, options)
    
    if not result.success:
        raise HTTPException(status_code=400, detail=result.error)
    
    return {
        "success": True,
        "skill": {
            "id": result.skill.id,
            "name": result.skill.name,
            "description": result.skill.description,
            "category": result.skill.category.value,
            "trigger_keywords": result.skill.trigger_keywords,
        } if result.skill else None,
        "analysis": {
            "doc_type": result.analysis.doc_type,
            "word_count": result.analysis.word_count,
            "complexity": result.analysis.complexity,
            "concepts": result.analysis.concepts[:10],
        } if result.analysis else None,
    }


@router.post("/batch-convert")
async def batch_convert_to_skill(
    document_ids: list[str],
    use_llm: bool = True,
    folder_id: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """批量将文档转换为 Skill
    
    Args:
        document_ids: 文档 ID 列表
        use_llm: 是否使用 LLM 生成
        folder_id: 目标文件夹 ID
    
    Returns:
        批量转换结果
    """
    options = ConvertOptions(
        use_llm=use_llm,
        folder_id=folder_id,
    )
    
    service = DocumentToSkillService(db)
    results = await service.batch_convert(document_ids, options)
    
    return {
        "total": len(results),
        "success_count": sum(1 for r in results if r.success),
        "results": [
            {
                "document_id": r.document_id,
                "success": r.success,
                "skill_id": r.skill.id if r.skill else None,
                "skill_name": r.skill.name if r.skill else None,
                "error": r.error,
            }
            for r in results
        ],
    }
